### DISCUSSION TITLE
sticky or follow mode for documentation

### DISCUSSION THREAD

**Discussion by nemethf** [company-mode/company-mode#1331] [Ideas]
`C-h` shows the documentation buffer temporary.  If the selection changes because of a `M-n`, then the user has to press `C-h` again to see the documentation for the new selection.  I wonder would it be acceptable to have a `C-u C-h` binding to a new follow-mode feature that doesn't hide the documentation buffer when the selection changes, but updates it with the new documentation.  And maybe a repeated `C-u C-h` could switch off this functionality.

(I think company-quickhelp provides something similar, but this feature request is about the dedicated documentation buffer.)

Thanks.

**Comment by dgutov** [comment]
Sure. Would you like to implement that?

**Reply by nemethf** [reply]
I'm not really familiar with the internals of company, but I've come up with the following rough patch.  It seems to work.  

Maybe it would be better to restore the window configuration when the user resets the new company--auto-update-doc to nil, but I don't know which configuration should be restored.  Or maybe it's OK to left the doc-buffer visible in this case.  What do you think?

Also shouldn't Emacs inform the user that "company-auto-update-documentation" has changed.  If yes, how? A simple `message` is likely to be overwritten, isn't it?

```diff --git a/company.el b/company.el
index a6fa086..0a234e7 100644
--- a/company.el
+++ b/company.el
@@ -1347,6 +1347,8 @@ company-call-frontends
              (error (error "Company: frontend %s error \"%s\" on command %s"
                            frontend (error-message-string err) command)))))
 
+(defvar company--auto-update-doc nil)
+
 (defun company-set-selection (selection &optional force-update)
   "Set SELECTION for company candidates.
 This will update `company-selection' and related variable.
@@ -1366,6 +1368,9 @@ company-set-selection
   (when (or force-update (not (equal selection company-selection)))
     (setq company-selection selection
           company-selection-changed t)
+    (when company--auto-update-doc
+      (company-call-backend 'doc-buffer
+                            (nth company-selection company-candidates)))
     (company-call-frontends 'update)))
 
 (defun company--group-lighter (candidate base)
@@ -2821,12 +2826,11 @@ company--unread-this-command-keys
                                  unread-command-events))
     (clear-this-command-keys t)))
 
-(defun company-show-doc-buffer ()
+(defun company-show-doc-buffer-0 ()
   "Temporarily show the documentation buffer for the selection."
   (interactive)
   (let ((other-window-scroll-buffer)
         (selection (or company-selection 0)))
-    (company--electric-do
       (let* ((selected (nth selection company-candidates))
              (doc-buffer (or (company-call-backend 'doc-buffer selected)
                              (user-error "No documentation available")))
@@ -2836,7 +2840,16 @@ company-show-doc-buffer
                 doc-buffer (car doc-buffer)))
         (setq other-window-scroll-buffer (get-buffer doc-buffer))
         (let ((win (display-buffer doc-buffer t)))
-          (set-window-start win (if start start (point-min))))))))
+          (set-window-start win (if start start (point-min)))))))
+
+(defun company-show-doc-buffer (&optional toggle-auto-update)
+  (interactive "P")
+  (when toggle-auto-update
+    (setq company--auto-update-doc (not company--auto-update-doc)))
+  (if company--auto-update-doc
+      (company-show-doc-buffer-0)
+    (company--electric-do
+      (company-show-doc-buffer-0))))
 (put 'company-show-doc-buffer 'company-keep t)
 
 (defun company-show-location ()
```

**Reply by dgutov** [reply]
I think `company-set-selection` should call `company-show-doc-buffer` rather than invoke `(company-call-backend 'doc-buffer` directly.

The implementation above may work, but only as long as the `doc-buffer` handler itself doesn't wrap itself in `save-excursion`, for instance.

**Reply by yugaego** [reply]
For reference: this was implemented in [PR #1334](https://github.com/company-mode/company-mode/pull/1334).

**Comment by aguynamedben** [comment]
That's sweet this was added. Does anybody know how to set this up with use-package? I can get the functionality working using a prefix argument, i.e. `C-u C-h`, but it's not working with this in my Emacs config using use-package:

```elisp
(use-package company
  :ensure
  :bind
  (:map company-active-map
              ("C-n". company-select-next)
              ("C-p". company-select-previous)
              ("M-<". company-select-first)
              ("M->". company-select-last))
  (:map company-mode-map
        ("<tab>". tab-indent-or-complete)
        ("TAB". tab-indent-or-complete))
  :custom
  (company-auto-update-doc t)
  :config
  (global-company-mode))
```

How can I set `company-auto-update-doc` to a non-nil value without having to hit the prefix argument every time?

**Reply by dgutov** [reply]
@aguynamedben It might depend on the version of Company you have installed. Did you get it from MELPA?
